exporter/opensearchexporter/model.go

```diff
@@ -0,0 +1,100 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package opensearchexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/opensearchexporter"
+
+import (
+	"bytes"
+	"encoding/json"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/opensearchexporter/internal/objmodel"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/traceutil"
+)
+
+type mappingModel interface {
+	encodeLog(pcommon.Resource, plog.LogRecord) ([]byte, error)
+	encodeSpan(pcommon.Resource, ptrace.Span) ([]byte, error)
+}
+
+// encodeModel tries to keep the event as close to the original open telemetry semantics as is.
+// No fields will be mapped by default.
+//
+// Field deduplication and dedotting of attributes is supported by the encodeModel.
+//
+// See: https://github.com/open-telemetry/oteps/blob/master/text/logs/0097-log-data-model.md
+type encodeModel struct {
+	dedup bool
+	dedot bool
+}
+
+const (
+	traceIDField   = "traceID"
+	spanIDField    = "spanID"
+	attributeField = "attribute"
+)
+
+func (m *encodeModel) encodeLog(resource pcommon.Resource, record plog.LogRecord) ([]byte, error) {
+	var document objmodel.Document
+	document.AddTimestamp("@timestamp", record.Timestamp()) // We use @timestamp in order to ensure that we can index if the default data stream logs template is used.
+	document.AddTraceID("TraceId", record.TraceID())
+	document.AddSpanID("SpanId", record.SpanID())
+	document.AddInt("TraceFlags", int64(record.Flags()))
+	document.AddString("SeverityText", record.SeverityText())
+	document.AddInt("SeverityNumber", int64(record.SeverityNumber()))
+	document.AddAttribute("Body", record.Body())
+	document.AddAttributes("Attributes", record.Attributes())
+	document.AddAttributes("Resource", resource.Attributes())
+
+	if m.dedup {
```

Comment:
Not possible with the way it was made, but should it be possible to have an object that is both dedup and dedot?